etcdserver/api/v3rpc/watch.go

```diff
@@ -184,6 +191,9 @@ func (sws *serverWatchStream) sendLoop() {
 			if err := sws.gRPCStream.Send(c); err != nil {
 				return
 			}
+			if c.Created && c.Canceled {
```

Comment:
I think the `if c.Canceled {}` below handles this case already. Regardless, a future watcher shouldn't `return` out of the send loop because that'll take down all the other watchers.